### ISSUE TITLE
Version 4 plans

### ISSUE THREAD

**Issue by Koenvh1** [Koenvh1/ets2-local-radio#154]
There are a couple of things I want to have done for a version 4 of Local Radio. The main overarching goal is to rewrite the core of Local Radio to make it work more reliably. Adding new functionality is a nice bonus if it can be done easily, but that is not the aim at the moment. Local Radio has changed a lot throughout the years, and thus certain design decisions that made sense at the time no longer do currently. This should also make it easier to modify Local Radio later.

This is the list currently:

- [ ] Adding support for one station being in multiple cities without showing up twice
- [ ] Replacing the PeerJS system with a system handled by the server
- [ ] Syncing the state between the web-frontend and the server
- [ ] Making the web-frontend reactive
- [ ] Using websockets for the communication between server and web-frontend
- [ ] Adding better favourites support
- [ ] Adding better proxy support
- [ ] Updating to a new version of the SCS SDK
- [ ] Switching to a different game overlay library
- [ ] Making it easy to switch between versions of the cities/stations
- [x] Updating the styling to match the logo's

If you have any wishes, feel free to let me know, although I will warn you in advance that I do not guarantee that I will implement it (I probably will not). However, I am looking for people who want to contribute to Local Radio. If you feel like doing some of the things above, or you propose something else, let me know.

(As for when this will be done: no idea, probably 2022 or later.)

**Comment by Escort95** [comment]
Step 2 is easy ;) just create a country, name it with radio name ex. "VOX FM" and then add cities where VOX FM is broadcasting in real life. I use this when i edit local radio for myself ;)   ( It's switches auto. with no pop-up / white noise)

**Comment by Koenvh1** [comment]
True, that is an ugly hack that works, but the entire point of version 4 is to get rid of those ugly hacks.

**Comment by MrJlien** [comment]
I would have an idea regarding the reception, the volume of the station would go and leave room for white noise when the amount of static reaches 1.00, and between 0.50 and 0.85 the volume weakens more and more with variations
So if I am in Spain, if I want to listen to a radio from Sweden the only thing I would hear would be only white noise
I think this is possible

I increased the size of the area so that the reception is ok on a radius of ~ 20km
SWR 3, for example, broadcast in southwestern Germany reaches as far as Bourges in France

Another thing that would be good, would be to add an option to display the radio and the signal at the bottom right (while remaining displayed) (which can be deactivated of course)
It could be great to know the reception in real time without looking on the page (or if it is possible to add it on the mobile interface)

Also enable / disable the fact that the station changes automatically

**Comment by Koenvh1** [comment]
The first thing is already included - the volume does start fluctuating. It hides stations out of reach by default, but you can turn them on again by clicking the "Toggle show all" button on the bottom. Mind you, your browser might start slowing down to a crawl if you do so.

No plans for a permanent overlay - it will most likely interfere with other interface elements. Might add the current signal strength to the mobile interface.

You can already change the fact that it changes automatically by setting a really large threshold in the config.

**Comment by MrJlien** [comment]
It would be cool if it is possible to completely mute the volume of the radio once amount of static reaches 1.00 (and between 0.90 and 1.00 a very weak signal audio for a more natural death to the signal)

I would love to be able to integrate the signal strength (the 5 bars) on my stream, (the mobile version or player page with signal strengh add on the top) would be perfect for that
Like this :
![image](https://user-images.githubusercontent.com/90514952/133308131-0e64709c-0352-4b13-828e-c5910eec2881.png)

Because on the main page as the radios change position when I drive it is impossible to frame the exact place (if you see what I mean)

**Comment by MrJlien** [comment]
On the current version, is it possible to modify the value so that at 1.00 the volume drops to 0?

I am really not familiar with the lines of codes ect

I see some interesting lines on "Main.js"
on "Local Radio \ web \ lib \ js"

![image](https://user-images.githubusercontent.com/90514952/133362602-5890a6b1-75af-4333-876d-27a27dcaf649.png)

If I want to keep it as currently from 0.45 (amount of static) it starts to fluctuate and that between 0.70 and 1.00 (amount of static) I want the volume to be much lower while still fluctuating automatically and past 1.00 there is no sound while keeping the white noise is it possible ?

"0.45" on the line corresponds to amount of static ?
![image](https://user-images.githubusercontent.com/90514952/133362676-6c110046-dee0-4482-b5b8-64eff76a073e.png)

**Comment by zflash** [comment]
I would like the following:

- Option to keep a station fixed when you drive out of range. In some counties in the US, radio stations are horrible. 
- Native VR support maybe? ;-)

**Comment by ryanmsj** [comment]
please add support for macOS

**Comment by Koenvh1** [comment]
@ryanmsj Not possible currently due to a limitation in ETS2/ATS.

**Comment by zflash** [comment]
I would love an option to lock a station regardless of my location at the time. There are some cool radio stations that I would like to listen to regardless of the county I'm in. As soon as I'm fed up with that station, simple click and I'm back at the location oriented stations.

**Comment by Koenvh1** [comment]
@zflash What do you mean exactly? A set number of stations always available, or ensuring that the current station does not switch? I have been thinking about the latter, but have not come up with a nice approachable way to implement it without causing confusion.

**Comment by zflash** [comment]
> @zflash What do you mean exactly? A set number of stations always available, or ensuring that the current station does not switch? I have been thinking about the latter, but have not come up with a nice approachable way to implement it without causing confusion.

For example. I'm a fan of rock and metal stations and there are some really good ones, so I favorite them. I would like to have an option to have an extra flag to specific stations that I can cycle through regardless of what state I'm in.  Basically I want to listen to a small station from Cali when I'm in Idaho.

**Comment by Koenvh1** [comment]
@zflash Ah okay. I believe you can do that already using the `global` country. It was initially thought of for something like TruckersFM, but it works with any audio stream.

**Comment by zflash** [comment]
@Koenvh1 How can I use the global country? Is that a setting in the ini? Would be a lot cooler that with a simple click I could just keep a station locked until I unlock it.

**Comment by Koenvh1** [comment]
It is something that can be done using `custom.js`, e.g.:
```js
stations["global"].push({
            name: "Station name",
            logo: "Logo URL",
            url: "Stream URL"
        });
```

The main issue with locking a couple of stations is that creating a list to pick from is difficult (as the number of available stations means that simply displaying them all will grind your web browser to a halt). Additionally, there is the question whether this should be ephemeral or a permanent thing that remains after reboots etc. Doing something permanently can already be done (admittedly, a button would be easier, but this is not too difficult either). Having something there temporarily is something I have been contemplating (e.g., you find a programme you like and want to hear the end of it).

**Comment by zflash** [comment]
@Koenvh1 A button for a temporary lock would be a welcome addition. If a permanent list of station you can rotate is a big issue, I would love just the lock function :)

**Comment by JulienEREZ** [comment]
For v4, it would be so nice if there was the possibility that the audio of the current radio changes from Stereo to Mono as soon as the whitenoise becomes audible in order to give a more realistic look

In the genre when the whitenoise reaches 0.10 it switches to mono

For my part, I made sure that there is gradually no longer any of the current station as soon as the whitenoise reaches 0.90

From 0.20 it begins to fluctuate, at 0.40 much more, 0.60 even more, then at 0.80 it is difficult to listen to it

instead of the whitenoise .mp3 file I put the link to a stream that broadcasts a real radio on the neighboring frequency (current station = 95.6, neighboring radio = 95.7)

Here video of the plugin in action:
https://www.youtube.com/watch?v=bDt6buU3GZw

Another thing that would be nice to do, when the truck is stationary, the random fluctuation effect stops, it would be active only when the truck is moving

**Comment by Koenvh1** [comment]
@JulienEREZ Not a bad idea - with the way it currently works it would be quite a feat to add though, but not impossible.

**Comment by WarLord211** [comment]
Hi i would have some QoL Ideas.

+more controllers. i would use buttons from the steeringwheel and from an self made buttonbox but atm its only 1 controller possible

+if the truck has no electric the radio turns not on or if it switched to off, it turns off to. - possible via telemetry.
 you get the coordinates, so like in the mobile advisor Stuff there is somerthing for electric and engine. 
 as a fun stuff if engine is turned on -> the radio is off for 1-2 sec. would be super immersive!

**Comment by relaxedcoder** [comment]
@Koenvh1 [QoL] Can we move to [SDK style projects](https://learn.microsoft.com/en-us/dotnet/core/project-sdk/overview) and [Central Package Management](https://learn.microsoft.com/en-us/nuget/consume-packages/central-package-management). Willing to work on this.
